Incorrect C# compilation of discriminators with underscores

If a discriminator has an underscore in it, Dafny generates invalid C# code. The method name will have that underscore replaced with two underscores, but any call to that method will have the underscore unchanged. This problem doesn't exist in Dafny release 3.2.0, but does exist in 3.3.0 and in the current version.

One can reproduce this and get the following error:
```
(1876,15): error CS1061: 'D' does not contain a definition for 'is_D_1' and no accessible extension method 'is_D_1' accepting a first argument of type 'D' could be found (are you missing a using directive or an assembly reference?)
```
by running Dafny on a file with the following contents:
```
module M {
  datatype D = D_1(i: int) | D_2(b: bool)

  method DoIt (d: D)
  {
    if (d.D_1?) {
      print d.i;
    }
    else {
      print d.b;
    }
  }
}
```

Running Dafny with `/spillTargetCode:3` further illuminates the issue.  One can see that the method is defined as:
```
    public bool is_D__1 { get { return this is D_D__1; } }
```
but the call to that method is:
```
      if ((d).is_D_1) {
```

--- a/Source/Dafny/Resolver.cs
+++ b/Source/Dafny/Resolver.cs
@@ -2109,7 +2109,7 @@ namespace Microsoft.Dafny {
 
               // create and add the query "method" (field, really)
               string queryName = ctor.Name + "?";
-              var query = new SpecialField(ctor.tok, queryName, SpecialField.ID.UseIdParam, "is_" + ctor.Name,
+              var query = new SpecialField(ctor.tok, queryName, SpecialField.ID.UseIdParam, "is_" + ctor.CompileName,
                 false, false, false, Type.Bool, null);
               query.InheritVisibility(dt);
               query.EnclosingClass = dt; // resolve here

--- a/Test/git-issues/git-issue-1607.dfy
+++ b/Test/git-issues/git-issue-1607.dfy
@@ -1,0 +1,26 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:cs "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:js "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:go "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:java "%s" >> "%t"
+// RUN: %diff "%s.expect" "%t"
+
+module M {
+  datatype D = D_1(a: bool) | D_2(b: bool)
+
+  method DoIt (d: D)
+  {
+    if (d.D_1?) {
+      print "1 ", d.a;
+    }
+    else {
+      print "2 ", d.b;
+    }
+  }
+}
+
+method Main(){
+  var x : M.D := M.D_2(true);
+  M.DoIt(x);
+  print "\n";
+}

--- a/Test/git-issues/git-issue-1607.dfy.expect
+++ b/Test/git-issues/git-issue-1607.dfy.expect
@@ -1,0 +1,14 @@
+
+Dafny program verifier finished with 2 verified, 0 errors
+
+Dafny program verifier did not attempt verification
+2 true
+
+Dafny program verifier did not attempt verification
+2 true
+
+Dafny program verifier did not attempt verification
+2 true
+
+Dafny program verifier did not attempt verification
+2 true

